pkg/alibabacloud/eni/node.go

```diff
@@ -0,0 +1,436 @@
+
+	n.manager.ForeachInstance(n.node.InstanceID(),
+		func(instanceID, interfaceID string, rev ipamTypes.InterfaceRevision) error {
+			e, ok := rev.Resource.(*eniTypes.ENI)
+			if ok {
+				resource.Status.AlibabaCloud.ENIs[interfaceID] = *e.DeepCopy()
+			}
+			return nil
+		})
+
+	return
+}
+
+// CreateInterface creates an additional interface with the instance and
+// attaches it to the instance as specified by the CiliumNode. neededAddresses
+// of secondary IPs are assigned to the interface up to the maximum number of
+// addresses as allowed by the instance.
+func (n *Node) CreateInterface(ctx context.Context, allocation *ipam.AllocationAction, scopedLog *logrus.Entry) (int, string, error) {
+	l, limitsAvailable := n.getLimits()
+	if !limitsAvailable {
+		return 0, errUnableToDetermineLimits, fmt.Errorf(errUnableToDetermineLimits)
+	}
+
+	n.mutex.RLock()
+	resource := *n.k8sObj
+	n.mutex.RUnlock()
+
+	// Must allocate secondary ENI IPs as needed, up to ENI instance limit - 1 (reserve 1 for primary IP)
+	toAllocate := math.IntMin(allocation.MaxIPsToAllocate, l.IPv4-1)
+	toAllocate = math.IntMin(maxENIIPCreate, toAllocate) // in first alloc no more than 9
+	// Validate whether request has already been fulfilled in the meantime
+	if toAllocate == 0 {
+		return 0, "", nil
+	}
+
+	bestSubnet := n.manager.FindOneVSwitch(resource.Spec.AlibabaCloud.VPCID, resource.Spec.AlibabaCloud.AvailabilityZone,
+		toAllocate, resource.Spec.AlibabaCloud.VSwitchTags)
+	if bestSubnet == nil {
+		return 0,
+			errUnableToFindSubnet,
+			fmt.Errorf(
+				"no matching vSwitch available for interface creation (VPC=%s AZ=%s SubnetTags=%s",
+				resource.Spec.AlibabaCloud.VPCID,
+				resource.Spec.AlibabaCloud.AvailabilityZone,
+				resource.Spec.AlibabaCloud.VSwitchTags,
+			)
+	}
+
+	securityGroupIDs, err := n.getSecurityGroupIDs(ctx, resource.Spec.AlibabaCloud)
+	if err != nil {
+		return 0,
+			errUnableToGetSecurityGroups,
+			fmt.Errorf("%s %s", errUnableToGetSecurityGroups, err)
+	}
+
+	scopedLog = scopedLog.WithFields(logrus.Fields{
+		"securityGroupIDs": securityGroupIDs,
+		"vSwitchID":        bestSubnet.ID,
+		"toAllocate":       toAllocate,
+	})
+	scopedLog.Info("No more IPs available, creating new ENI")
+
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+	index, err := n.allocENIIndex()
+	if err != nil {
+		scopedLog.WithField("instanceID", n.node.InstanceID()).Error(err)
+		return 0, "", err
+	}
+	eniID, eni, err := n.manager.api.CreateNetworkInterface(ctx, toAllocate, bestSubnet.ID, securityGroupIDs,
+		utils.FillTagWithENIIndex(map[string]string{}, index))
+	if err != nil {
+		return 0, errUnableToCreateENI, fmt.Errorf("%s %s", errUnableToCreateENI, err)
+	}
+
+	scopedLog = scopedLog.WithField(fieldENIID, eniID)
+	scopedLog.Info("Created new ENI")
+
+	err = n.manager.api.AttachNetworkInterface(ctx, n.node.InstanceID(), eniID)
+	if err != nil {
+		return 0, errUnableToAttachENI, fmt.Errorf("%s %s", errUnableToAttachENI, err)
+	}
+	time.Sleep(2 * time.Second)
```

Comment:
I guess I didn't spot this last time. We should not have sleeps in the production code, besides in tests. Is this necessary? If waiting is necessary, can you utilize Go channels?

Reply:
not quite necessary .  Added  only bacause AttachNetworkInterface is async will take a fell seconds to really success .